types/src/v0/impls/stake_table.rs

```diff
@@ -264,128 +287,222 @@ impl Membership<SeqTypes> for StaticCommittee {
             .collect();
 
         // Index the stake table by public key
-        let indexed_da_stake_table: BTreeMap<PubKey, _> = da_members
+        let indexed_da_members: BTreeMap<PubKey, _> = da_members
             .iter()
             .map(|entry| (PubKey::public_key(entry), entry.clone()))
             .collect();
 
-        Self {
+        let members = Committee {
             eligible_leaders,
-            stake_table: HashSet::from_iter(members),
-            da_stake_table: HashSet::from_iter(da_members),
             indexed_stake_table,
-            indexed_da_stake_table,
-            _epoch_size: 12, // TODO get the real number from config (I think)
-            _provider: L1Client::http(Url::from_str("http:://ab.b").unwrap()),
-            _contract_address: None,
+            indexed_da_members,
+        };
+
+        let mut map = HashMap::new();
+        map.insert(Epoch::genesis(), members);
+
+        Self {
+            state: Arc::new(RwLock::new(map)),
+            _epoch_size: 12,
+            l1_client: L1Client::http(Url::from_str("http:://ab.b").unwrap()),
         }
     }
+
     /// Get the stake table for the current view
-    fn stake_table(&self, _epoch: Epoch) -> Vec<StakeTableEntry<PubKey>> {
-        self.stake_table.clone().into_iter().collect()
+    fn stake_table(&self, epoch: Epoch) -> Vec<StakeTableEntry<PubKey>> {
+        let state = self.state.read_blocking();
+
+        match state.get(&epoch) {
+            Some(st) => st.indexed_stake_table.clone().into_values().collect(),
+            None => {
+                let stake_tables = self.l1_client.stake_table(&epoch);
```

Comment:
I don't think we are waiting are doing any long waits here, but as best practice it may be nice to release the lock at the beginning of the `None` branch. Is this possible?

Reply:
I think 79c1617cf4c96f7440b8aa419dd035e1ac21666d will do this. Depending on if NLL does what I think it does.